### ISSUE TITLE
Suggest use of unwrap_or_default

### ISSUE THREAD

**Issue by ndmitchell** [rust-lang/rust-clippy#7250]
### What it does

Suggest using `Option::unwrap_or_default` when `Option::unwrap_or_else` is called using a function that creates the default.

### Categories (optional)

- Kind: style

Makes the code shorter and more idiomatic.

### Drawbacks

None.

### Example

```rust
let x : Option<_> = ...
x.unwrap_or_else(Vec::new)
```

Could be written as:

```rust
let x : Option<_> = ...
x.unwrap_or_default()
```

Requires an understanding of what closures produce default values - e.g. `Vec::default`, `Default::default`, `Vec::new`, `|| Vec::new()`.

**Comment by pmnoxx** [comment]
+1, I see a lot of code in our code base that's like
`tmp.get(arg).cloned().unwrap_or(0u64)`
or
`dict.get(arg).unwrap_or(HashMap::new())`

**Comment by giraffate** [comment]
Oh, we've already have the [`unwrap_or_else_default`](https://rust-lang.github.io/rust-clippy/master/index.html#unwrap_or_else_default) lint.

**Comment by pmnoxx** [comment]
> Oh, we've already have the [`unwrap_or_else_default`](https://rust-lang.github.io/rust-clippy/master/index.html#unwrap_or_else_default) lint.

Yes, I saw that. Though it only works when used with `<TYPE>::default`.
Some values, like `0u64 == u64::default`, but the link will not work for them.

**Comment by pmnoxx** [comment]
@giraffate Maybe, there should be another int, let's call it 

name: `use_default`.
Example:
```
func(0u64)
```
Warnings message:
`0u64` could be replaced with `u64::default`

This way, we would convert values to default, and then `unwrap_or_else_default` would do the rest of the work.

**Comment by giraffate** [comment]
> Yes, I saw that. Though it only works when used with <TYPE>::default.
> For example:
> unwrap_or(u64::default) -> works
> unwrap_or(0u64) -> doesn't.

IMO [`or_fun_call`](https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call) should cover these cases.

> This way, we would convert values to default, and then unwrap_or_else_default would do the rest of the work.

I don't want to make Clippy run multiple times. I want it to suggest the desired code once.

**Comment by pmnoxx** [comment]
> > Yes, I saw that. Though it only works when used with ::default.
> > For example:
> > unwrap_or(u64::default) -> works
> > unwrap_or(0u64) -> doesn't.
> 
> IMO [`or_fun_call`](https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call) should cover these cases.
> 
> > This way, we would convert values to default, and then unwrap_or_else_default would do the rest of the work.
> 
> I don't want to make Clippy run multiple times. I want it to suggest the desired code once.



> > Yes, I saw that. Though it only works when used with ::default.
> > For example:
> > unwrap_or(u64::default) -> works
> > unwrap_or(0u64) -> doesn't.
> 
> IMO [`or_fun_call`](https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call) should cover these cases.
> 
> > This way, we would convert values to default, and then unwrap_or_else_default would do the rest of the work.
> 
> I don't want to make Clippy run multiple times. I want it to suggest the desired code once.

I'm wondering whenever something like this would work: https://github.com/rust-lang/rust-clippy/pull/8163

**Comment by giraffate** [comment]
https://github.com/rust-lang/rust-clippy/pull/8163 fixes this, so I'm closing this.
